### ISSUE TITLE
proposed Publication DOI validation for incongruency detection

### ISSUE THREAD

**Issue by adf-ncgr** [LegumeFederation/datastore#30]
we've been pretty good as a whole about supplying this in a form that is resolvable using the
DOI resolution service at http://dx.doi.org/<Dataset DOI>

I just modified 393b1fb the README template to suggest that DOIs supplied in forms that do not resolve should be considered incorrect. Suggesting now that we put some code behind this assertion.

@ctcncgr could either do this himself, or use it as an opportunity to train someone else how to add teeth to the evolving standards...

**Comment by ctcncgr** [comment]
http://www.doi.org/factsheets/DOIProxy.html#rest-api

Looks straightforward.  @adf-ncgr Are the DOIs retrieved from the README?

**Comment by adf-ncgr** [comment]
Yes, DOIs are actually two places  in the README.<key>.md (I had overlooked one earlier):
#### Dataset DOI 
 
none

#### Publication DOI 
 
none

**Comment by ctcncgr** [comment]
Ok, gotcha.  I will check them if available.

**Comment by cann0010** [comment]
Agreed that this is a good idea. Probably should wait to close this until we have implemented this somehow - e.g. reviewing all instances of the DOIs in the READMEs, and documenting in the protocols. If I recall, few datasets actually have a DOI, but most of the datasets have a publication - and should therefore have a publication DOI.

**Comment by adf-ncgr** [comment]
if you want to review, take a look at files on lis-*:
~adf/md-publication_doi
~adf/md-dataset_doi

generated with cmds like:
[adf@lis-stage /usr/local/www/data]$ for f in */*/*/README.????.md; do echo -n "$f "; ./private/scripts/metadata_md_to_csv.pl $f | grep '^dataset_doi'; done > ~/md-dataset_doi

note that I do get a number of perl complaints from this, such as:
Can't use string ("") as a HASH ref while "strict refs" in use at ./private/scripts/metadata_md_to_csv.pl line 163, <$IN_FH> line 137.
Use of uninitialized value within %dataset_H in numeric eq (==) at ./private/scripts/metadata_md_to_csv.pl line 88, <$IN_FH> line 102.

not sure if this means the parser is too frail or the files are too rowdy, but I have ignored it for now.

**Comment by ctcncgr** [comment]
Gotcha.  I'll make sure this is documented from a programmatic standpoint.  

A new "validate_doi" method will be added to the Detector class.  It will make a callout to https://doi.org/api/handles/.  The value given should be &lt;handle>/&lt;suffix> or possibly just handle, but I think there should be a suffix.

Response Codes

  * 1 : Success. (HTTP 200 OK)
  * 2 : Error. Something unexpected went wrong during handle resolution. (HTTP 500 Internal Server Error)
  * 100 : Handle Not Found. (HTTP 404 Not Found)
  * 200 : Values Not Found. The handle exists but has no values (or no values according to the types and indices specified). (HTTP 200 OK)

Basically if its not a 1.  I think it should cry foul.

**Comment by adf-ncgr** [comment]
sounds like you already know more about their API than I do, but I agree that for our purposes it should be :
<handle>/<suffix>

**Comment by adf-ncgr** [comment]
Yes, your suggestion of what is sensible is also what I believe (and seems to be what is most common in use already in the files, at least for the publication DOIs)

**Comment by ctcncgr** [comment]
Invalid DOIs from directories containing canonical data:
  * Cicer_arietinum/CDCFrontier.gnm1.GkHc/README.GkHc.md
    * `//dx.doi.org/10.5524/100076 INVALID`
  * Cicer_arietinum/CDCFrontier.gnm1.ann1.nRhs/README.nRhs.md
    * `//dx.doi.org/10.5524/100076 INVALID`
  * Lupinus_angustifolius/Tanjil.gnm1.Qq0N/README.Qq0N.md
    * `#### GenBank Accession INVALID`
  * Lupinus_angustifolius/Tanjil.gnm1.ann1.nnV9/README.nnV9.md
    * `#### GenBank Accession INVALID`
  * Phaseolus_vulgaris/G19833.gnm1.zBnF/README.zBnF.md
    * `#### GenBank Accession INVALID`
  * Phaseolus_vulgaris/G19833.gnm1.ann1.pScz/README.pScz.md
    * `#### GenBank Accession INVALID`
  * Phaseolus_vulgaris/G19833.gnm2.fC0g/README.fC0g.md
    * `#### Dataset Release Date INVALID`
  * Phaseolus_vulgaris/G19833.gnm2.ann1.PB8d/README.PB8d.md
    * `#### Dataset Release Date INVALID`
    * `#### GenBank Accession INVALID`
  * Trifolium_pratense/MilvusB.gnm2.gNmT/README.gNmT.md
    * `#### GenBank Accession INVALID`

Most of these are either formatted wrong or missing 'none' for their value

**Comment by jd-campbell** [comment]
Fixed. Added <handle> or missing 'none' value.

**Comment by adf-ncgr** [comment]
I'm not sure we're on the same page about this; I just did some spot-checking and not clear to me that things were resolved according to our interpretation of the intention of these metadata slots. But can wait to see what the validator says. In the meantime, it looks like the README file for 
Cicer_arietinum/CDCFrontier.gnm1.GkHc/README.GkHc.md
has gone AWOL from the datastore. Any idea what happened to it?

**Comment by jd-campbell** [comment]
I was just working on that README: Cicer_arietinum/CDCFrontier.gnm1.GkHc/README.GkHc.md
